beacon_node/beacon_chain/src/validator_custody.rs

```diff
@@ -0,0 +1,403 @@
+use std::{
+    collections::{BTreeMap, HashMap},
+    sync::atomic::{AtomicU64, Ordering},
+};
+
+use parking_lot::RwLock;
+
+use ssz_derive::{Decode, Encode};
+use tokio::sync::broadcast::{channel, Receiver, Sender};
+use types::{ChainSpec, Epoch, EthSpec, Slot};
+
+const CHANNEL_CAPACITY: usize = 10;
+
+/// TODO(pawan): this currently just registers increases in validator count.
+/// Does not handle decreasing validator counts
+#[derive(Default, Debug)]
+struct ValidatorRegistrations {
+    /// Set of all validators that is registered to this node along with its effective balance
+    ///
+    /// Key is validator index and value is effective_balance.
+    validators: HashMap<usize, u64>,
+    /// Maintains the validator custody requirement at a given epoch.
+    ///
+    /// Note: Only stores the epoch value when there's a change in custody requirement.
+    /// So if epoch 10 and 11 has the same custody requirement, only 10 is stored.
+    epoch_validator_custody_requirements: BTreeMap<Epoch, u64>,
+}
+
+impl ValidatorRegistrations {
+    /// Returns the validator custody requirement at the latest epoch.
+    pub fn latest_validator_custody_requirement(&self) -> Option<u64> {
+        self.epoch_validator_custody_requirements
+            .last_key_value()
+            .map(|(_, v)| *v)
+    }
+
+    /// Returns the latest epoch at which the validator count changed.
+    #[allow(dead_code)]
+    pub fn latest_epoch(&self) -> Option<Epoch> {
+        self.epoch_validator_custody_requirements
+            .last_key_value()
+            .map(|(k, _)| *k)
+    }
+
+    /// Register a new validator index and updates the list of validators if required.
+    pub fn register_validator<E: EthSpec>(
+        &mut self,
+        validator_index: usize,
+        effective_balance: u64,
+        slot: Slot,
+        spec: &ChainSpec,
+    ) {
+        let epoch = slot.epoch(E::slots_per_epoch());
```

Comment:
The benefit of tracking the epoch in the updates is to uphold the following invariant

> A given block will have the same CGC during its lifetime 

To prevent situations where we receive the block from gossip with a CGC of 8 and the moment of importing turns out we are at 16, failing import. But the columns have already been gossiped (and ignored by us because we were CGC 8). In that case we have to trigger lookup sync to fetch the remaining columns and recover. It may work but it feels messy. Instead we can schedule the update for the next epoch to prevent this edge case.